LANGUAGE_SDK_SPECIFICATION.md

```diff
@@ -0,0 +1,1325 @@
+4. The user can submit the callback ID to external systems
+
+#### 7.2.2 Wait for Callback
+
+The SDK SHOULD provide a pattern where:
+
+1. A CALLBACK operation is created
+2. A user-provided function receives the callback ID
+3. The function submits the ID to external systems
+4. The operation waits for external completion
+5. Retry logic can be applied to the submission function
+
+### 7.3 Promise Combinators
+
+#### 7.3.1 Purpose
+
+Coordinate multiple durable promises in familiar patterns.
+
+#### 7.3.2 Supported Combinators
+
+The SDK SHOULD provide equivalents to common promise patterns:
+
+- **all**: Wait for all promises to complete successfully
+- **allSettled**: Wait for all promises to settle (success or failure)
+- **race**: Return the first promise to settle
+- **any**: Return the first promise to succeed
+
+#### 7.3.3 Implementation Considerations
+
+The SDK SHOULD:
+
+1. Document that combinators work with already-started operations
+2. Recommend map/parallel for controlled concurrency and durability
+3. Implement combinators using CONTEXT operations for tracking
+4. Support proper error propagation and result collection
+
+## 8. Error Handling
+
+### 8.1 Error Classification
+
+The SDK MUST distinguish between:
+
+1. **Runtime Errors**: Errors in the execution environment
+   - Prefixed with `Runtime.`, `Sandbox.`, or `Extension.`
+   - Automatically retried by Lambda (up to 3 immediate retries)
+   - Exception: `Sandbox.Timedout` is treated as handler error
+
+2. **Handler Errors**: Errors thrown by user code
+   - All other error types
+   - Retried with exponential backoff over several hours
+   - Expected to be caught and handled by SDK
+
+### 8.2 Error Recovery
+
+#### 8.2.1 Retryable Errors
+
+The SDK SHOULD treat errors as retryable by default and:
+
+1. Allow the function to be re-invoked by Lambda
+2. Rely on replay to skip completed operations
+3. Continue from the point of failure
+
+#### 8.2.2 Terminal Errors
+
+The SDK MUST return `FAILED` status for terminal errors:
+
+1. User explicitly requests failure
+2. Retry budget is exhausted
+3. Operation reaches terminal failed state
+4. Unrecoverable validation or serialization errors
+
+### 8.3 Error Propagation
+
+#### 8.3.1 Step Errors
+
+When a step fails:
+
+1. The SDK SHOULD check the retry strategy
+2. If retry is requested, checkpoint with `RETRY` action
+3. If retry is exhausted, checkpoint with `FAIL` action and propagate error
+
+#### 8.3.2 Context Errors
+
+When a child operation fails within a context:
+
+1. The SDK SHOULD allow context-level error handling
+2. The SDK MAY cancel other pending operations in the context
+3. The SDK MUST checkpoint the context failure if not recovered
+
+#### 8.3.3 Error Objects
+
+The SDK MUST provide error objects with:
+
+- `ErrorType`: String identifying the error class
+- `ErrorMessage`: Human-readable error description
+
+The SDK SHOULD include:
+
+- `StackTrace`: Array of stack frame strings
```

Comment:
There's also a 4th field `ErrorData` which can be used for additional machine-readable error information (e.g. serialized JSON)

https://docs.aws.amazon.com/lambda/latest/api/API_ErrorObject.html